enterprise/backend/src/metabase_enterprise/sandbox/models/params/field_values.clj

```diff
@@ -1,21 +1,66 @@
 (ns metabase-enterprise.sandbox.models.params.field-values
   (:require [metabase-enterprise.sandbox.api.table :as table]
+            [metabase-enterprise.sandbox.models.group-table-access-policy :refer [GroupTableAccessPolicy]]
+            [metabase-enterprise.sandbox.query-processor.middleware.row-level-restrictions :as row-level-restrictions]
             [metabase.api.common :as api]
-            [metabase.models.field :as field :refer [Field]]
+            [metabase.mbql.util :as mbql.u]
+            [metabase.models :refer [Field PermissionsGroupMembership User]]
+            [metabase.models.field :as field]
             [metabase.models.field-values :as field-values]
             [metabase.models.params.field-values :as params.field-values]
             [metabase.public-settings.premium-features :refer [defenterprise]]
             [toucan.db :as db]))
 
 (comment api/keep-me)
 
-(defn- field-is-sandboxed?
+(defn field-is-sandboxed?
+  "Check if a field is sandboxed."
   [{:keys [table], :as field}]
   ;; slight optimization: for the `field-id->field-values` version we can batched hydrate `:table` to avoid having to
   ;; make a bunch of calls to fetch Table. For `get-or-create-field-values` we don't hydrate `:table` so we can fall
   ;; back to fetching it manually with `field/table`
   (table/only-segmented-perms? (or table (field/table field))))
 
+(defn- table-id->gtap
+  "Find the GTAP for current user that apply to table `table-id`."
+  [table-id]
+  (let [group-ids (db/select-field :group_id PermissionsGroupMembership :user_id api/*current-user-id*)
+        gtaps     (db/select GroupTableAccessPolicy
+                             :group_id [:in group-ids]
+                             :table_id table-id)]
+    (when gtaps
+      (row-level-restrictions/assert-one-gtap-per-table gtaps)
+      ;; there shold be only one gtap per table and we only need one table here
+      ;; see docs in [[metabase.models.permissions]] for more info
+      (first gtaps))))
+
+(defn- field->gtap-attributes-for-current-user
+  "Returns the gtap attributes for current user that applied to `field`.
+
+  The gtap-attributes is a list with 2 elements:
+  1. card-id - for GTAP that use a saved question
+  2. a map:
+    - with key is the user-attribute applied to `field-id`
+    - value is the user-attribute of current user corresponding to the key
+
+  For example we have an GTAP rules
+  {:card_id 1
+   :attribute_remappings {\"State\" [:dimension [:field 3 nil]]}}
+
+  And users with login-attributes {\"State\" \"CA\"}
+
+  ;; (field-id->gtap-attributes-for-current-user (Field 3))
+  ;; -> [1, {\"State\" \"CA\"}]"
+  [{:keys [table_id id] :as _field}]
+  (when-let [gtap (table-id->gtap table_id)]
+    (let [login-attributes (or (:login_attributes @api/*current-user*)
```

Comment:
In what circumstances would `(:login_attributes @api/*current-user*)` be nil but we would find attributes in the db with `(db/select-one-field :login_attributes User :id api/*current-user-id*)`?